editor/components/inserter/menu.js

```diff
@@ -12,10 +12,161 @@ import Dashicon from 'components/dashicon';
 class InserterMenu extends wp.element.Component {
 	constructor() {
 		super( ...arguments );
+		this.blockTypes = wp.blocks.getBlocks();
+		this.categories = wp.blocks.getCategories();
+		this.nodes = {};
 		this.state = {
-			filterValue: ''
+			filterValue: '',
+			focusedElementRef: null
 		};
 		this.filter = this.filter.bind( this );
+		this.isShownBlock = this.isShownBlock.bind( this );
+		this.changeMenuSelection = this.changeMenuSelection.bind( this );
+		this.setSearchFocus = this.setSearchFocus.bind( this );
+		this.bindReferenceNode = this.bindReferenceNode.bind( this );
+		this.onKeyDown = this.onKeyDown.bind( this );
+		this.listVisibleBlocksByCategory = this.listVisibleBlocksByCategory.bind( this );
+	}
+
+	componentDidMount() {
+		document.addEventListener( 'keydown', this.onKeyDown );
+	}
+
+	componentWillUnmount() {
+		document.removeEventListener( 'keydown', this.onKeyDown );
+	}
+
+	isShown( filterValue, string ) {
+		return string.toLowerCase().indexOf( filterValue.toLowerCase() ) !== -1;
+	}
+
+	isShownBlock( filterValue, block ) {
+		return this.isShown( filterValue, block.title );
+	}
+
+	visibleFocusables( nodes, filterValue ) {
+		const focusables = [];
+
+		for ( const refName in nodes ) {
+			if ( this.isShown( filterValue, refName ) && 'search' !== refName ) {
```

Comment:
I think we should have a unique `isShownBlock` method to check for visibility, because we may want to change the way we filter the blocks into a more complex way (keywords, title...). We already talked about moving this filtering function into the `blocks` module.